src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskStatusUpdateProcessorImpl.scala

```diff
@@ -1,133 +1,85 @@
-  //scalastyle:off cyclomatic.complexity method.length
-  def publish(incoming: TaskStatusObservables.TaskStatusUpdate): Future[Unit] = incoming match {
-    case TaskStatusUpdate(timestamp, taskId, WithMesosStatus(status)) =>
-      val appId = taskIdUtil.appId(taskId)
+  private[this] val publishFutureTimer: Timer =
+    metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, "publishFuture"))
+  private[this] val fetchTaskTimer: Timer = metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, "fetchTask"))
+  private[this] val getAppTimer: Timer = metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, "getApp"))
 
-      // forward health changes to the health check manager
-      val maybeTask = taskTracker.fetchTask(appId, taskId.getValue)
-      for (marathonTask <- maybeTask)
-        healthCheckManager.update(status, Timestamp(marathonTask.getVersion))
+  private[this] val stepTimers: Map[String, Timer] = steps.map { step =>
+    step.name -> metrics.timer(metrics.name(MetricPrefixes.SERVICE, getClass, s"step-${step.name}"))
+  }.toMap
 
-      import org.apache.mesos.Protos.TaskState._
+  log.info("Started status update processor with steps:\n{}", steps.map(step => s"* ${step.name}").mkString("\n"))
 
-      import scala.concurrent.ExecutionContext.Implicits.global
-      val eventuallyTracked = status.getState match {
-        case TASK_ERROR | TASK_FAILED | TASK_FINISHED | TASK_KILLED | TASK_LOST =>
-          // Remove from our internal list
-          taskTracker.terminated(appId, taskId.getValue).flatMap { taskOption =>
-            rateLimit(appId, status, taskOption).map { _ =>
-              schedulerActor ! ScaleApp(appId)
+  override def publish(status: TaskStatus): Future[Unit] = publishFutureTimer.timeFuture {
+    val now = clock.now()
+    val taskId = status.getTaskId
+    val appId = taskIdUtil.appId(taskId)
 
-              taskOption match {
-                case Some(task) => postEvent(status, task)
-                case None       => log.warn(s"Task not found. Do not post event for '{}'", taskId.getValue)
-              }
-            }
-          }
+    val maybeTask = fetchTaskTimer(taskTracker.fetchTask(appId, taskId.getValue))
 
-        case TASK_RUNNING if !maybeTask.exists(_.hasStartedAt) => // staged, not running
-          taskTracker.running(appId, status)
-            .map { task => postEvent(status, task) }
-            .recover {
-              case NonFatal(t) =>
-                log.warn(s"Task could not be saved. Do not post event for '${taskId.getValue}'", t)
-                driverOpt.foreach(_.killTask(status.getTaskId))
-            }
-
-        case TASK_STAGING if !taskTracker.contains(appId) =>
-          log.warn(s"Received status update for unknown app $appId, killing task ${status.getTaskId}")
-          driverOpt.foreach(_.killTask(status.getTaskId))
-          Future.successful(())
+    val eventuallyMaybeApp = maybeTask match {
```

Comment:
You are right. It is only used in `NotifyRateLimiterStepImpl` if a task has finished but not killed.

good catch!